ocean_lib/models/test/test_datatoken3.py

```diff
@@ -0,0 +1,297 @@
+
+S_PER_MIN = 60
+S_PER_HOUR = 60 * 60
+
+
+@enforce_types
+def test_main():
+    # ======================================================================
+    # SETUP
+    config, ocean, OCEAN, accts = _setup()
+    [deployer, opf, predictoor0, predictoor1, trader, rando, DT_treasurer] = accts
+    predictoors = [predictoor0, predictoor1]
+
+    # convenience functions
+    OCEAN_bal = ConvClass(OCEAN).fromWei_balanceOf
+    OCEAN_approved = ConvClass(OCEAN).fromWei_approved
+
+    # ======================================================================
+    # PARAMETERS
+    s_per_block = 2  # depends on the chain
+    s_per_epoch = 5 * S_PER_MIN
+    s_per_subscription = 24 * S_PER_HOUR
+    min_predns_for_payout = 3  # ideally, 100+
+    stake_token = OCEAN
+    stakes = [2.0, 1.0]  # Stake per predictoor. In OCEAN
+
+    n_DTs = 100.0  # num DTs for sale
+    DT_price = 10.0  # denominated in OCEAN
+
+    max_n_predns = min_predns_for_payout  # stop loop after this many predn's
+
+    x = ocean.data_nft_factory.getCurrentTemplateCount()
+
+    # ======================================================================
+    # DEPLOY DATATOKEN & EXCHANGE CONTRACT
+    data_nft = ocean.data_nft_factory.create({"from": opf}, "DN", "DN")
+
+    # since data_nft.create_datatoken does not allow us to send all the arrays, we have to do it manually
+    # DT = data_nft.create_datatoken({"from": opf}, "DT", "DT", template_index=3)
+    initial_list = data_nft.getTokensList()
+    data_nft.createERC20(
+        3,
+        ["DT3", "DT3"],
+        [opf.address, opf.address, opf.address, OCEAN.address, OCEAN.address],
+        [MAX_UINT256, 0, s_per_block, s_per_epoch, s_per_subscription, 30],
+        [],
+        {"from": opf},
+    )
+    new_elements = [
+        item for item in data_nft.getTokensList() if item not in initial_list
+    ]
+    assert len(new_elements) == 1, "new datatoken has no address"
+    DT = DatatokenBase.get_typed(config, new_elements[0])
+
+    # Mixed sol+py can't have DT to hold OCEAN, so have a different account
+    # - When we convert to Solidity, change to: DT_treasurer = DT
+    DT.treasurer = DT_treasurer
+
+    # post DTs for sale
+    DT.setup_exchange({"from": opf}, to_wei(DT_price))
+    DT.approve(DT.exchange.address, to_wei(n_DTs), {"from": opf})
+
+    # ======================================================================
+    # TRADER BUYS SUBSCRIPTION TO AGG PREDVALS (do every 24h)
+
+    # trader buys 1 DT with OCEAN
+    exchange = DT.get_exchanges()[0]
+    OCEAN_needed = from_wei(exchange.BT_needed(to_wei(1), consume_market_fee=0))
+    OCEAN.approve(exchange.address, to_wei(OCEAN_needed), {"from": trader})
+
+    consume_market_fee_addr = ZERO_ADDRESS
+    consume_market_fee = 0
+    tx = exchange.buy_DT(
+        datatoken_amt=to_wei(1),
+        consume_market_fee_addr=consume_market_fee_addr,
+        consume_market_fee=consume_market_fee,
+        tx_dict={"from": trader},
+    )
```

Comment:
Can simplify, because in `buy_DT()` the default consume market fee is zero. [Ref](https://github.com/oceanprotocol/ocean.py/blob/a8a8a94fd97a149629cb69df1701108d3a637b41/ocean_lib/models/fixed_rate_exchange.py#L309)

And, you don't need the return `tx`.

Therefore it will look like:
```python
    exchange.buy_DT(datatoken_amt=to_wei(1), tx_dict={"from": trader})
```